File: android/app/src/main/java/com/americangroupllc/buddyplay/games/chess/ChessViewModel.kt
package com.americangroupllc.buddyplay.games.chess

import androidx.lifecycle.ViewModel
import com.americangroupllc.buddyplay.core.domain.ChessMove
import com.americangroupllc.buddyplay.core.domain.ChessPieceKind
import com.americangroupllc.buddyplay.core.domain.ChessRules
import com.americangroupllc.buddyplay.core.domain.ChessSquare
import com.americangroupllc.buddyplay.core.domain.ChessState
import com.americangroupllc.buddyplay.core.domain.GameStateReducer
import com.americangroupllc.buddyplay.core.models.Peer
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

class ChessViewModel(host: Peer, guest: Peer) : ViewModel() {

    val host: Peer = host
    val guest: Peer = guest

    var state: ChessState by mutableStateOf(ChessRules.initialState(host, guest))
        private set

    var selected: ChessSquare? by mutableStateOf(null)
        private set
    var legalDestinations: Set<ChessSquare> by mutableStateOf(emptySet())
        private set
    var lastError: String? by mutableStateOf(null)
        private set

    val isInCheck: Boolean get() = ChessRules.isInCheck(state, state.sideToMove)

    fun tap(square: ChessSquare) {
        val from = selected
        if (from != null) {
            if (square in legalDestinations) {
                commit(ChessMove(from, square, promotion = defaultPromotion(from, square)))
            } else {
                val piece = state.board[square]
                if (piece != null && piece.color == state.sideToMove) select(square)
                else clearSelection()
            }
        } else {
            select(square)
        }
    }

    private fun select(square: ChessSquare) {
        val piece = state.board[square]
        if (piece == null || piece.color != state.sideToMove) {
            clearSelection()
            return
        }
        selected = square
        legalDestinations = ChessRules.legalMoves(state, piece.color)
            .filter { it.from == square }
            .map { it.to }
            .toSet()
    }

    private fun clearSelection() {
        selected = null
        legalDestinations = emptySet()
    }

    private fun commit(move: ChessMove) {
        try {
            val step = ChessRules.reduce(state, move)
            state = step.state
            clearSelection()
        } catch (e: GameStateReducer.Error) {
            lastError = e.message
            clearSelection()
        }
    }

    private fun defaultPromotion(from: ChessSquare, to: ChessSquare): ChessPieceKind? {
        val p = state.board[from] ?: return null
        if (p.kind != ChessPieceKind.PAWN) return null
        val promoRank = if (p.color == com.americangroupllc.buddyplay.core.domain.ChessColor.WHITE) 7 else 0
        return if (to.rank == promoRank) ChessPieceKind.QUEEN else null
    }
}

File: apps/BuddyPlay/ios/BuddyPlay/Features/Games/Chess/ChessViewModel.swift
import Foundation
import Combine
import BuddyCore

/// View model for `ChessScreen`. Holds the current `ChessState` and routes
/// taps through `ChessRules.reduce`.
///
/// In a 2-device session the host's view model owns the authoritative
/// state; the guest's view model receives state snapshots from the
/// connectivity layer and renders them. v1 ships local-only play first;
/// the wire-up to `GameSessionService` lands together with the
/// connectivity adapters.
@MainActor
final class ChessViewModel: ObservableObject {

    @Published private(set) var state: ChessState
    @Published var selected: ChessSquare?
    @Published var legalDestinations: Set<ChessSquare> = []
    @Published var lastError: String?

    let host: Peer
    let guest: Peer

    init(host: Peer, guest: Peer) {
        self.host = host
        self.guest = guest
        self.state = ChessRules.initialState(host: host, guest: guest)
    }

    /// Tap handler. First tap selects a piece (and computes legal
    /// destinations); second tap on a legal destination commits the move.
    func tap(_ square: ChessSquare) {
        if let from = selected {
            if legalDestinations.contains(square) {
                commit(ChessMove(from: from, to: square, promotion: defaultPromotion(from: from, to: square)))
            } else if let piece = state.board[square], piece.color == state.sideToMove {
                select(square)
            } else {
                clearSelection()
            }
        } else {
            select(square)
        }
    }

    private func select(_ square: ChessSquare) {
        guard let piece = state.board[square], piece.color == state.sideToMove else {
            clearSelection()
            return
        }
        selected = square
        let allLegal = ChessRules.legalMoves(in: state, for: piece.color)
        legalDestinations = Set(allLegal.filter { $0.from == square }.map { $0.to })
    }

    private func clearSelection() {
        selected = nil
        legalDestinations = []
    }

    private func commit(_ move: ChessMove) {
        do {
            let step = try ChessRules.reduce(state, input: move)
            state = step.state
            clearSelection()
            SfxService.shared.playMove()
            if step.outcome != nil { SfxService.shared.playWin() }
        } catch {
            lastError = "\(error)"
            clearSelection()
        }
    }

    /// v1: always promote to queen automatically. v1.1 adds a picker.
    private func defaultPromotion(from: ChessSquare, to: ChessSquare) -> ChessPiece.Kind? {
        guard let p = state.board[from], p.kind == .pawn else { return nil }
        let promoRank = p.color == .white ? 7 : 0
        return to.rank == promoRank ? .queen : nil
    }

    var isInCheck: Bool { ChessRules.isInCheck(state, color: state.sideToMove) }
    var currentTurn: UUID? { ChessRules.currentTurn(state) }
}
